daemon/src/setup_contract.rs

```diff
@@ -113,18 +99,204 @@ pub async fn new(
+        .try_into_msg2()?;
+
+    tracing::debug_span!("Merge lock PSBTs").in_scope(|| {
+        signed_lock_tx
+            .combine(msg2.signed_lock)
+            .context("Failed to merge lock PSBTs")
+    })?;
+
+    let cets = extract_counterparty_adaptor_sig(
+        &params,
+        verified.commit_tx.clone(),
+        verified.commit_desc.clone(),
+        verified.own_cets,
+        msg1.cets,
+    )
+    .await?;
+
+    // TODO: Remove send- and receiving ACK messages once we are able to handle incomplete DLC
+    // monitoring
+    sink.send(SetupMsg::Msg3(Msg3))
+        .instrument(tracing::debug_span!("Send Msg3"))
+        .await
+        .context("Failed to send Msg3")?;
+    let _ = stream
+        .next()
+        .timeout(CONTRACT_SETUP_MSG_TIMEOUT, stream_next_span)
+        .await
+        .with_context(|| {
+            format!(
+                "Expected Msg3 within {} seconds",
+                CONTRACT_SETUP_MSG_TIMEOUT.as_secs()
+            )
+        })?
+        .context("Empty stream instead of Msg3")?
+        .try_into_msg3()?;
+
+    Ok(Dlc {
+        identity: key_pairs.identity.private,
+        identity_counterparty: params.counterparty.identity_pk,
+        revocation: key_pairs.revoke.private,
+        revocation_pk_counterparty: params.counterparty_punish.revocation_pk,
+        publish: key_pairs.publish.private,
+        publish_pk_counterparty: params.counterparty_punish.publish_pk,
+        maker_address: params.maker().address.clone(),
+        taker_address: params.taker().address.clone(),
+        lock: (signed_lock_tx.extract_tx(), verified.lock_desc),
+        commit: (verified.commit_tx, msg1.commit, verified.commit_desc),
+        cets,
+        refund: (verified.refund_tx, msg1.refund),
+        maker_lock_amount: params.maker().lock_amount,
+        taker_lock_amount: params.taker().lock_amount,
+        revoked_commit: Vec::new(),
+        settlement_event_id,
+        refund_timelock: setup_params.refund_timelock,
+    })
+}
+
+fn stream_next_span() -> tracing::Span {
+    tracing::debug_span!("Receive setup message")
+}
+
+#[derive(Copy, Clone)]
+pub struct KeyPair {
+    private: SecretKey,
+    public: PublicKey,
+}
+
+impl From<(SecretKey, PublicKey)> for KeyPair {
+    fn from((private, public): (SecretKey, PublicKey)) -> Self {
+        KeyPair { private, public }
+    }
+}
+
+#[derive(Copy, Clone)]
+struct KeyPairs {
+    identity: KeyPair,
+    revoke: KeyPair,
+    publish: KeyPair,
+}
 
+#[instrument(name = "Generate own params", skip_all, err)]
+async fn own_setup_params(
+    build_party_params_channel: MessageChannel<wallet::BuildPartyParams, Result<PartyParams>>,
+    setup_params: SetupParams,
+) -> Result<(PartyParams, PunishParams, KeyPairs)> {
+    let key_pairs = KeyPairs {
+        identity: keypair::new(&mut rand::thread_rng()).into(),
+        revoke: keypair::new(&mut rand::thread_rng()).into(),
+        publish: keypair::new(&mut rand::thread_rng()).into(),
+    };
```

Comment:
nit: this could be `KeyPairs` new()  or default() function instead? otherwise it's pretty hard to see the purpose of the abstraction.

Reply:
That's true. Currently it's only here to group everything together to avoid having to pass 3 args to anything that needs keys